sparkly/instant_testing.py

```diff
@@ -0,0 +1,159 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import json
+import logging
+import os
+import signal
+import tempfile
+
+from py4j.java_gateway import java_import
+from pyspark import SparkContext
+from pyspark.java_gateway import launch_gateway
+
+
+logger = logging.getLogger(__name__)
+
+
+class InstantTesting(object):
+    """The set of tools to run tests using Spark Context running in the background.
+
+    Implementation:
+        We create a lock file that will contain Python gateway port (exposed by JVM).
+
+        On the first run:
+            - initialise Spark Context as usual;
+            - write Python gateway port to the lock file;
+            - fork current process.
+
+        On the second run:
+            - connect to the background JVM process using Python gateway port from the lock file;
+            - recover Spark Context from JVM.
+    """
+    LOCK_FILE_PATH = os.path.join(tempfile.gettempdir(), 'sparkly_instant_testing_lock')
+
+    @classmethod
+    def activate(cls):
+        """Activate instant testing mode."""
+        if os.path.exists(cls.LOCK_FILE_PATH):
+            logger.error('Instant testing mode is already activate, deactivate it first.')
+        else:
+            with open(cls.LOCK_FILE_PATH, 'w'):
+                logger.info('Instant testing mode has been activated.')
+
+    @classmethod
+    def deactivate(cls):
+        """Deactivate instance testing mode."""
+        if not os.path.exists(cls.LOCK_FILE_PATH):
+            logger.error('Instant testing mode is not activated, activate it first.')
+        else:
+            try:
+                with open(cls.LOCK_FILE_PATH) as lock:
+                    state = lock.read()
+                    if state:
+                        session_pid = json.loads(state)['session_pid']
+                        try:
+                            os.kill(session_pid, signal.SIGTERM)
+                        except OSError:
+                            logger.exception(
+                                'Can not kill background SparkContext (pid %d)', session_pid,
+                            )
+                        else:
+                            logger.info(
+                                'Killed background SparkContext (pid %d)', session_pid,
+                            )
+            finally:
+                try:
+                    os.remove(cls.LOCK_FILE_PATH)
+                except OSError:
+                    logger.exception('Can not remove lock file')
```

Comment:
I know it's unlikely case, but let's add lock path to the message for easier debugging